File: src/main/java/io/github/prathamjainone/productivity/users/service/UserService.java
package io.github.prathamjainone.productivity.users.service;

import io.github.prathamjainone.productivity.users.exceptions.UserNotFoundException;
import io.github.prathamjainone.productivity.users.exceptions.UserValidationException;
import io.github.prathamjainone.productivity.users.model.User;

import java.util.List;

/**
 * Interface for user management services.
 * Defines operations for creating, reading, updating, and deleting user data.
 * Implementations of this interface should handle the business logic for user operations
 * and interact with persistent storage or in-memory data.
 *
 * <p>
 * The methods throw specific exceptions when an operation cannot be performed, such as:
 * </p>
 *
 * <ul>
 *     <li>{@link UserValidationException} for invalid user data.</li>
 *     <li>{@link UserNotFoundException} for non-existent users.</li>
 * </ul>
 */
public interface UserService {

    /**
     * Creates a new user.
     *
     * @param user the {@link User} object to create
     * @return the created {@link User}
     * @throws UserValidationException if the provided user data is invalid
     */
    User createUser(User user) throws UserValidationException;

    /**
     * Retrieves all users.
     *
     * @return a list of all {@link User} objects
     * @throws UserNotFoundException if no users are found
     */
    List<User> getAllUsers() throws UserNotFoundException;

    /**
     * Retrieves a user by their unique ID.
     *
     * @param id the unique identifier of the user
     * @return the {@link User} with the given ID
     * @throws UserNotFoundException if no user with the given ID exists
     */
    User getUserById(long id) throws UserNotFoundException;

    /**
     * Retrieves a user by their username.
     *
     * @param username the username of the user
     * @return the {@link User} with the given username
     * @throws UserNotFoundException if no user with the given username exists
     */
    User getUserByUsername(String username) throws UserNotFoundException;

    /**
     * Updates the password of an existing user.
     *
     * @param id the unique identifier of the user
     * @param newPassword the new password for the user
     * @return the updated {@link User} entity
     * @throws UserValidationException if the new password is invalid
     */
    User updateUserPassword(long id, String newPassword) throws UserValidationException;

    /**
     * Updates the username of an existing user.
     *
     * @param id the unique identifier of the user
     * @param newUsername the new username for the user
     * @return the updated {@link User} entity
     * @throws UserValidationException if the new username is invalid
     */
    User updateUsername(long id, String newUsername) throws UserValidationException;

    /**
     * Updates the email address of an existing user.
     *
     * @param id the unique identifier of the user
     * @param newEmail the new email address for the user
     * @return the updated {@link User} entity
     * @throws UserValidationException if the new email is invalid
     */
    User updateEmail(long id, String newEmail) throws UserValidationException;

    /**
     * Updates the full name of an existing user.
     *
     * @param id the unique identifier of the user
     * @param newFullName the new full name for the user
     * @return the updated {@link User} entity
     * @throws UserValidationException if the new full name is invalid
     */
    User updateFullName(long id, String newFullName) throws UserValidationException;

    /**
     * Deletes a user by their unique ID.
     *
     * @param id the unique identifier of the user
     * @return {@code true} if the user was deleted successfully, {@code false} otherwise
     * @throws UserNotFoundException if no user with the given ID exists
     */
    boolean deleteUser(long id) throws UserNotFoundException;
}

File: frontend/src/services/userService.js
import api from './api';

/**
 * Service for interacting with user-related API endpoints
 */
const userService = {
  /**
   * Get all users
   * @returns {Promise} Promise with the list of users
   */
  getAllUsers: async () => {
    try {
      const response = await api.get('/users');
      return response.data;
    } catch (error) {
      console.error('Error fetching users:', error);
      throw error;
    }
  },

  /**
   * Get a user by ID
   * @param {number} id - User ID
   * @returns {Promise} Promise with the user data
   */
  getUserById: async (id) => {
    try {
      const response = await api.get(`/users/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching user with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Get a user by username
   * @param {string} username - Username
   * @returns {Promise} Promise with the user data
   */
  getUserByUsername: async (username) => {
    try {
      const response = await api.get('/users/byUsername', {
        params: { username }
      });
      return response.data;
    } catch (error) {
      console.error(`Error fetching user with username ${username}:`, error);
      throw error;
    }
  },

  /**
   * Create a new user
   * @param {Object} user - User data
   * @returns {Promise} Promise with the created user
   */
  createUser: async (user) => {
    try {
      const response = await api.post('/users', user);
      return response.data;
    } catch (error) {
      console.error('Error creating user:', error);
      throw error;
    }
  },

  /**
   * Update a user's full name
   * @param {number} id - User ID
   * @param {string} fullName - New full name
   * @returns {Promise} Promise with the updated user
   */
  updateFullName: async (id, fullName) => {
    try {
      const response = await api.put(`/users/${id}/fullName`, null, {
        params: { newFullName: fullName }
      });
      return response.data;
    } catch (error) {
      console.error(`Error updating user full name for ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Update a user's email
   * @param {number} id - User ID
   * @param {string} email - New email
   * @returns {Promise} Promise with the updated user
   */
  updateEmail: async (id, email) => {
    try {
      const response = await api.put(`/users/${id}/email`, null, {
        params: { newEmail: email }
      });
      return response.data;
    } catch (error) {
      console.error(`Error updating user email for ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Update a user's username
   * @param {number} id - User ID
   * @param {string} username - New username
   * @returns {Promise} Promise with the updated user
   */
  updateUsername: async (id, username) => {
    try {
      const response = await api.put(`/users/${id}/username`, null, {
        params: { newUsername: username }
      });
      return response.data;
    } catch (error) {
      console.error(`Error updating user username for ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Update a user's password
   * @param {number} id - User ID
   * @param {string} currentPassword - Current password
   * @param {string} newPassword - New password
   * @returns {Promise} Promise with the updated user
   */
  updatePassword: async (id, currentPassword, newPassword) => {
    try {
      const response = await api.put(`/users/${id}/password`, {
        currentPassword,
        newPassword
      });
      return response.data;
    } catch (error) {
      console.error(`Error updating user password for ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Update a user's profile information
   * @param {number} id - User ID
   * @param {Object} userData - User data to update
   * @returns {Promise} Promise with the updated user
   */
  updateUser: async (id, userData) => {
    try {
      const response = await api.put(`/users/${id}`, userData);
      return response.data;
    } catch (error) {
      console.error(`Error updating user with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a user
   * @param {number} id - User ID
   * @returns {Promise} Promise with the deletion result
   */
  deleteUser: async (id) => {
    try {
      const response = await api.delete(`/users/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting user with ID ${id}:`, error);
      throw error;
    }
  }
};

export default userService; 